### ISSUE TITLE
GPU Implementation of sort! and partialsort!

### ISSUE THREAD

**Issue by btmit** [JuliaGPU/CuArrays.jl#669]
It would be great if there was an implementation of sort! and partialsort! for CuArrays.

A simple (1D array, radix-2) native Julia implementation would be really helpful as a good starting point.

Other issues have considered wrapping Thrust or CUB #525.

This should enable a number of higher level functions, such as the ability to generate statistics about data (median, quartiles, etc) on the GPU. 

A few (possibly) helpful links:
https://www.cse.chalmers.se/~tsigas/papers/GPU-Quicksort-jea.pdf
https://onlinelibrary.wiley.com/doi/full/10.1002/cpe.3611
https://sourceforge.net/projects/cuda-quicksort/

**Comment by wongalvis14** [comment]
I looked into a couple parallel sorting algorithms each with their own setbacks:

Bitonic sort - for arrays with lengths not a power of 2, we need to fill in dummy values to make it so, the asymptotic complexity is the same, but the overhead and extra memory will be large;
or we can devise a non-recursive version that works based on https://www.inf.hs-flensburg.de/lang/algorithmen/sortieren/bitonic/oddn.htm, if an efficient solution exists

Merge sort - extra memory, more work for each thread (binary search) on the merge step, special handling for arrays with lengths not a power of 2

Radix sort - handling for negative numbers

Quick sort - bad worst case complexity

**Comment by maleadt** [comment]
I had tried a quicksort implementation to test dynamic parallelism, but it's slow: https://gist.github.com/maleadt/08bd84906382ed8b9c57e154356f6a0b

**Comment by wongalvis14** [comment]
CUDA's quicksort examples use dynamic parallelism, my experience with dynamic parallelism was very... not ideal... but I did come across a paper recently that showed a speedup of dynamic parallelism over device-host sync+dispatch

**Comment by wongalvis14** [comment]
I benchmarked my implementation of bitonic sort and @maleadt 's quicksort 

It seems that bitonic sort is a better baseline to build upon?

```
julia> @benchmark test_bitonic_sort(1000000)
BenchmarkTools.Trial: 
  memory estimate:  253.53 KiB
  allocs estimate:  8858
  --------------
  minimum time:     2.811 ms (0.00% GC)
  median time:      29.530 ms (0.00% GC)
  mean time:        29.380 ms (0.00% GC)
  maximum time:     30.135 ms (0.00% GC)
  --------------
  samples:          171
  evals/sample:     1

julia> @benchmark test_quicksort(1000000)
BenchmarkTools.Trial: 
  memory estimate:  1.53 KiB
  allocs estimate:  57
  --------------
  minimum time:     995.534 ms (0.00% GC)
  median time:      1.536 s (0.00% GC)
  mean time:        1.578 s (0.00% GC)
  maximum time:     2.243 s (0.00% GC)
  --------------
  samples:          4
  evals/sample:     1

julia> @benchmark test_quicksort(100000)
BenchmarkTools.Trial: 
  memory estimate:  1.53 KiB
  allocs estimate:  57
  --------------
  minimum time:     202.019 μs (0.00% GC)
  median time:      43.582 ms (0.00% GC)
  mean time:        52.244 ms (0.00% GC)
  maximum time:     148.446 ms (0.00% GC)
  --------------
  samples:          96
  evals/sample:     1
```

**Comment by maleadt** [comment]
> It seems that bitonic sort is a better baseline to build upon?

Definitely, but the pow2 requirement is problematic. Padding the array is a no-go, and using sentinel values to denote out-of-bounds elements is generally not possible either.

**Comment by wongalvis14** [comment]
For comparison, this is the standard Julia CPU sort
```
julia> @benchmark test_sort_cpu(1000000)
BenchmarkTools.Trial: 
  memory estimate:  15.26 MiB
  allocs estimate:  4
  --------------
  minimum time:     131.498 ms (0.00% GC)
  median time:      134.018 ms (0.00% GC)
  mean time:        136.183 ms (1.00% GC)
  maximum time:     150.196 ms (1.20% GC)
  --------------
  samples:          37
  evals/sample:     1
```

> the pow2 requirement is problematic

Indeed, in the beginning I was in search of a generic solution for all array lengths. Until I saw NVIDIA's implementation, in which they added a memory buffer (I learnt it from them). I realized this is probably the "best solution" so far. On the other hand, I think a lot of times we sort arrays that are either pow2 or pow10 which would be relatively close to a pow2 anyway.
